apisix/plugins/rocketmq-logger.lua

```diff
@@ -0,0 +1,256 @@
+        max_retry_count = {type = "integer", minimum = 0, default = 0},
+        retry_delay = {type = "integer", minimum = 0, default = 1},
+        buffer_duration = {type = "integer", minimum = 1, default = 60},
+        inactive_timeout = {type = "integer", minimum = 1, default = 5},
+        include_req_body = {type = "boolean", default = false},
+        include_req_body_expr = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "array",
+                items = {
+                    type = "string"
+                }
+            }
+        },
+        include_resp_body = {type = "boolean", default = false},
+        include_resp_body_expr = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "array",
+                items = {
+                    type = "string"
+                }
+            }
+        },
+    },
+    required = {"nameserver_list", "topic"}
+}
+
+local metadata_schema = {
+    type = "object",
+    properties = {
+        log_format = log_util.metadata_schema_log_format,
+    },
+}
+
+local _M = {
+    version = 0.1,
+    priority = 402,
+    name = plugin_name,
+    schema = schema,
+    metadata_schema = metadata_schema,
+}
+
+
+function _M.check_schema(conf, schema_type)
+    if schema_type == core.schema.TYPE_METADATA then
+        return core.schema.check(metadata_schema, conf)
+    end
+
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return nil, err
+    end
+    return log_util.check_log_schema(conf)
+end
+
+
+-- remove stale objects from the memory after timer expires
+local function remove_stale_objects(premature)
+    if premature then
+        return
+    end
+
+    for key, batch in ipairs(buffers) do
+        if #batch.entry_buffer.entries == 0 and #batch.batch_to_process == 0 then
+            core.log.warn("removing batch processor stale object, conf: ",
+                          core.json.delay_encode(key))
+            buffers[key] = nil
+        end
+    end
+
+    stale_timer_running = false
+end
+
+
+local function create_producer(nameserver_list, producer_config)
+    core.log.info("create new rocketmq producer instance")
+    local prod = producer.new(nameserver_list, "apisixLogProducer")
+    if producer_config.use_tls then
+        prod:setUseTLS(true)
+    end
+    if producer_config.access_key ~= '' then
+        local aclHook = acl_rpchook.new(producer_config.access_key, producer_config.secret_key)
+        prod:addRPCHook(aclHook)
+    end
+    prod:setTimeout(producer_config.timeout)
+    return prod
+end
+
+
+local function send_rocketmq_data(conf, log_message, prod)
+    local result, err = prod:send(conf.topic, log_message, conf.tag, conf.key)
+    if not result then
+        return false, "failed to send data to rocketmq topic: " .. err ..
+                ", nameserver_list: " .. core.json.encode(conf.nameserver_list)
```

Comment:
Can we get the nameserver for the current error?

Reply:
Sorry, I dont understand what you mean. the nameserver is add to err and test case will verify the error log if send fail

Reply:
if we have more nameservers in nameserver_list, one of them is unavailable and this message is sent to the unavailable nameserver, can the log here only print out the unavailable nameserver?

Ok, if this is not convenient to do in this PR, we can optimise it later.

Reply:
Which nameserver is error can be returned from the lua-resty-rocketmq library, We can create an issue there.